plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/utils/SourceElementExtensions.java

```diff
@@ -0,0 +1,318 @@
+		Pair<String, VariableEnvironmentElement> key = Pair.of("collectLocalArguments", element);
+		return cache.get(key, element.eResource(), () -> doCollectLocalArguments(element));
+	}
+
+	private List<IdentifiableElement> doCollectVisibleIdentifiableElements(VariableEnvironmentElement element) {
+		List<IdentifiableElement> result = new ArrayList<>();
+		if (element instanceof IdentifiableElement) {
+			result.add((IdentifiableElement) element);
+
+		} else if (element instanceof FunctionExpression && ((FunctionExpression) element).getName() != null) {
+			collectVisibleTypedElement((FunctionExpression) element, result);
+		}
+
+		List<IdentifiableElement> visibleIdentifiableElements = doCollectVisibleIdentifiableElements(element,
+				element, true);
+
+		result.addAll(visibleIdentifiableElements);
+
+		return result;
+	}
+
+	private List<IdentifiableElement> doCollectLocalArguments(VariableEnvironmentElement element) {
+		List<IdentifiableElement> result = new ArrayList<>();
+		if (element instanceof FunctionOrFieldAccessor) {
+			if (!(element instanceof ArrowFunction)) {
+				result.add(((FunctionOrFieldAccessor) element).getLocalArgumentsVariable());
+			}
+		}
+		return result;
+	}
+
+	/**
+	 * Walk through all contents list of the given element. If the entry in the list is a N4ClassDefinition, a
+	 * N4EnumDeclaration, a ExportedVariableDeclaration or a FunctionDefinition its children are walked, too. If the
+	 * entry in the list is a IdentifiableElement this is also collected. If the entry in the list is a before mentioned
+	 * element or a Expression the iterator is pruned as it is clear that below these elements no further visible
+	 * identifiable elements can be found.
+	 */
+	private List<IdentifiableElement> doCollectVisibleIdentifiableElements(VariableEnvironmentElement start,
+			EObject element, boolean includeBlockScopedElements) {
+
+		if (includeBlockScopedElements) {
+			return doCollectVisibleIdentifiableElementsUncached(start, element, includeBlockScopedElements);
+
+		} else {
+			return cache.get(Pair.of("doCollectVisibleIdentifiableElements?includeBlockScopedElements=false", element),
+					element.eResource(), () -> {
+						return doCollectVisibleIdentifiableElementsUncached(start, element, includeBlockScopedElements);
+					});
+		}
+	}
+
+	private List<IdentifiableElement> doCollectVisibleIdentifiableElementsUncached(VariableEnvironmentElement start,
+			EObject element, boolean includeBlockScopedElements) {
+
+		List<IdentifiableElement> result = new ArrayList<>();
+		TreeIterator<EObject> allContents = element.eAllContents();
+		InternalCFEMapper veeSwitch = new InternalCFEMapper(start, includeBlockScopedElements, allContents, result);
+		while (allContents.hasNext()) {
+			EObject next = allContents.next();
+			veeSwitch.doSwitch(next);
+		}
+		return result;
+	}
+
+	private class InternalCFEMapper extends N4JSSwitch<Boolean> {
```

Comment:
I guess CFE means "control flow element"? If so, the class name should be updated.